src/engine/TransitivePathBinSearch.h

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Johannes Herrmann (johannes.r.herrmann(at)gmail.com)
+
+#pragma once
+
+#include <iterator>
+#include <memory>
+
+#include "engine/Operation.h"
+#include "engine/QueryExecutionTree.h"
+#include "engine/TransitivePathImpl.h"
+#include "engine/idTable/IdTable.h"
+
+/**
+ * @class BinSearchMap
+ * @brief This struct represents a simple Binary Search Map. Given an Id, the
+ * BinSearchMap can return a list (i.e. span) of successor Ids.
+ *
+ * It is expected that the two input spans startIds_ and targetIds_ are sorted
+ * first by start id and then by target id.
+ * Example:
+ *
+ * startId | targetId
+ * ------------------
+ *       1 |        1
+ *       1 |        2
+ *       2 |        4
+ *       3 |        2
+ *       3 |        4
+ *       3 |        6
+ *       5 |        2
+ *       5 |        6
+ *
+ */
+struct BinSearchMap {
+  std::span<const Id> startIds_;
+  std::span<const Id> targetIds_;
+
+  /**
+   * @brief Return the successors for the given id.
+   * The successors are all target ids, where the corresponding start id is
+   * equal to the given id `node`.
+   *
+   * @param node The input id, which will be looked up in startIds_
+   * @return A std::span<Id>, which consists of all targetIds_ where
+   * startIds_ == node.
+   */
+  auto successors(const Id node) const {
+    auto range = std::ranges::equal_range(startIds_, node);
+
+    auto startIndex = std::distance(startIds_.begin(), range.begin());
+
+    return targetIds_.subspan(startIndex, range.size());
```

Comment:
```suggestion
    return std::ranges::equal_range(startIds_, node);
```
Should work, as everything is templated, you don't need to know the type, and it just needs to be iterable.

Reply:
No, this does not work. The result of the code you provided are the start ids, which match the node and not the target ids.

Example:
startIds_: 1, 1, 2
targetIds_: 2, 3, 4

`equal_range(startIds_, 1)` returns `{1, 1}`, but the correct result would be `{2, 3}`.

Another option would be to write a wrapper for `std::pair<Id, Id>` with a custom comparison function. And it would be necessary to convert the two `std::span<Id>` into a single `std::span<std::pair<Id, Id>>`. But I would argue that this solution is unnecessarily complex, considering the same effect can be achieved with three lines of code.